### ISSUE TITLE
LBFGS accuracy difference CPU/GPU

### ISSUE THREAD

**Issue by joacorapela** [pytorch/pytorch#111508]
### 🐛 Describe the bug

I am running a [complex minimization algorithm](https://github.com/joacorapela/svGPFA) using LBFGS in PyTorch. I have a method ``model.to(device)`` that moves all model variables to ``device``. If I run the optimization with the model on ``cpu`` (i.e., calling ``model.to(torch.device("cpu"))``) the algorithm converges to a relatively large value. But if I run the same optimization algorithm, with the same data, but with the model on gpu (i.e., calling ``model.to(torch.device("cuda:0"))``) the algorithm converges to a much smaller value. I would appreciate any hint on why this could be happening.

The algorithm runs slower in GPU than in CPU, but this may be a consequence or running for more iterations in GPU.

Sorry I cannot post a minimal working example, but I could not reproduce the problem with simpler optimizations. If it helps, I can create a Google Colab notebook replicating this problem.

### Versions

$ python collect_env.py
Collecting environment information...
PyTorch version: 2.0.0+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.26.0
Libc version: glibc-2.31

Python version: 3.9.16 | packaged by conda-forge | (main, Feb  1 2023, 21:39:03)  [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-109-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: Quadro P5000
GPU 1: NVIDIA GeForce RTX 2080

Nvidia driver version: 525.125.06
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          40
On-line CPU(s) list:             0-39
Thread(s) per core:              2
Core(s) per socket:              10
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           63
Model name:                      Intel(R) Xeon(R) CPU E5-2650 v3 @ 2.30GHz
Stepping:                        2
CPU MHz:                         1199.283
CPU max MHz:                     3000.0000
CPU min MHz:                     1200.0000
BogoMIPS:                        4594.42
Virtualization:                  VT-x
L1d cache:                       640 KiB
L1i cache:                       640 KiB
L2 cache:                        5 MiB
L3 cache:                        50 MiB
NUMA node0 CPU(s):               0-9,20-29
NUMA node1 CPU(s):               10-19,30-39
Vulnerability Itlb multihit:     KVM: Mitigation: Split huge pages
Vulnerability L1tf:              Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb invpcid_single pti intel_ppin ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm xsaveopt cqm_llc cqm_occup_llc dtherm ida arat pln pts md_clear flush_l1d

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] numpy==1.24.3
[pip3] torch==2.0.0
[pip3] triton==2.0.0
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] torch                     2.0.0                    pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi

**Comment by colesbury** [comment]
Outside of a few simple operations (addition, multiplication) you should not expect the same results on CPU and GPU. And if the steps are not identical, iterative algorithms quickly diverge. i.e., even if the first step is similar on CPU and GPU, it's pretty normal for the final result to be different.

**Comment by joacorapela** [comment]
Thanks for your reply @colesbury . The striking fact is that for different initial parameters of the optimisation algorithm, and even for different training datasets, the algorithm always achieves a better result in GPU than in CPU. I will create a notebook reproducing this problem, so that my problem can be replicated.

**Comment by joacorapela** [comment]
Hello @colesbury. To facilitate the reproduction of this problem, I created a [Google Colab notebook](https://colab.research.google.com/github/joacorapela/svGPFA_simulations/blob/master/bugReport/plotGPU_LBFGS_bug.ipynb) that runs the same optimization algorithm, with the same data and parameters, in CPU and GPU. This code demonstrates that LBFGS optimizations generate better estimates (i.e., larger lower bounds) when the device of the tensors is set to cuda:0 than to cpu.                 
                                                                                                                                                                                                                                                       
I agree with your previous point that we should not expect the same results in GPU and CPU, especially in stochastic optimization algorithms. However, I cannot understand why optimization results are always far superior in GPU than in CPU. The superiority of GPU optimizations holds for different initial conditions and also for different data sets (not shown in this example, that uses only one dataset).                                                                                   
                                                                                                                                                                                                                                                       
Could you help me understand the reason for this problem? Please let me know if I could provide further information.                                                                                                                                   
                                                                                                                                                                                                                                                       
May the implementation of LBFGS in CPU be different from that in GPU?                                                                                                                                                                                  
                                                                                                                                                                                                                                                       
I would have guessed that the implementation in GPU could perform worse than that in CPU, as resources in GPUs are more limited than those in CPUs. But here I am observing the opposite case.                                                         
                                                                                                                                                                                                                                                       
I will next prepare another Google Colab notebook demonstrating this same problem but with a completely different dataset.                                                                                                                             
                                                                                                                                                                                                                                                       
It seems that the CPU implementation always gets trapped in a local minima with a lower bound of 20670, but the CUDA implementation escapes from this local minima and achieves much larger, and varied, lower bounds.                                    
                                                                                                                                                                                                                                                       
I would greatly appreciate any hint.                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                       
Thanks, Joaquin

**Comment by albanD** [comment]
Some details are here as well btw: https://pytorch.org/docs/stable/notes/randomness.html
